### PR TITLE
[autoscaler] Fix update/terminate race condition

### LINKED ISSUE #14996: [Kubernetes][autoscaler] More problems during scale down
### What is the problem?

We're still experiencing concurrency issues during scale down on K8s.
Logs: https://controlc.com/e40990dd

Will use comments on this issue to log investigation.

Discussion context: 
https://discuss.ray.io/t/stuck-trying-to-take-down-workers/1309/21

Previous issue:
https://github.com/ray-project/ray/issues/14812

Previous PR:
https://github.com/ray-project/ray/pull/14976

*Ray version and other system information (Python version, TensorFlow version, OS):*

Autoscaler using recent master.

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

It seems it's possible for a node to be terminated while a NodeUpdaterThread is trying to update it.
This causes the update to fail. The autoscaler then repeatedly tries to terminate the node again which leads to an error. The autoscaler continues to try to delete the non-existent node until the autoscaler exceeds `max_failures` and stops.

This PR
(1) Modifies autoscaler logic to avoid the race condition
(2) Modifies `kubernetes_node_provider.terminate_node()` so that it handles the 404 error if it tries to delete a missing pod. 

See https://github.com/ray-project/ray/issues/14996 .

## Related issue number

Closes https://github.com/ray-project/ray/issues/14996

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

Open to any testing suggestions / any ideas on how to replicate the error in https://github.com/ray-project/ray/issues/14996 .

### CODE REVIEW

**Comment by wuisawesome** [comment]
Is this a race condition?

If so, shouldn't we fix it with a trying to terminate all the nodes and catching any exceptions (ideally terminating a node twice shouldn't be an error right?). 

Also, wouldn't we need to do this in all the other places that we terminate nodes?

**Comment by DmitriGekhtman** [comment]
> Is this a race condition?
> 
> If so, shouldn't we fix it with a trying to terminate all the nodes and catching any exceptions (ideally terminating a node twice shouldn't be an error right?).
> 
> Also, wouldn't we need to do this in all the other places that we terminate nodes?

Different node providers produce different errors when you try to terminate a node that's not there, so like you said we would have to catch any exception each time we call `provider.terminate_nodes` ---
```
try:
    provider.terminate_nodes
except Exception as e:
    blah
```
which I think is not so great -- would ideally want to catch a specific exception.

**Comment by DmitriGekhtman** [comment]
Some additional error handling could make things safer, though -- in addition to the change in this PR, we could replace each autoscaler call to `self.provider.non_terminated_nodes` to a call to a method like this.

```
def _terminate_nodes(self, node_ids):
    try:
        self.provider.terminate_nodes(node_ids)
    except Exception:
        logger.warning("Failed to terminate nodes. Trying again.")
        active_nodes = self.workers()
        for node_id in node_ids:
            if node_id in active_nodes:
                self.provider.terminate_node(node_id)
            else:
                logger.warning("Node {} has already been terminated.".format(node_id))
```
and something similar for calls to `self.provider.terminate_node`.

Idea is that node termination would be safer, but also there's a warning logged somewhere that you've tried to delete a non-existent node, which is not a thing that should happen. 

@wuisawesome What do you think?

**Review by AmeerHajAli — CHANGES_REQUESTED** [review]
Why are we terminating these nodes in first place? These nodes should not exist in general (how did we end up with more nodes than max_workers?)
What is the problem with terminating updating nodes?

**Comment by DmitriGekhtman** [comment]
> @DmitriGekhtman This function (`def _terminate_nodes(self, node_ids):`) looks reasonable.
> 
> My concern is that this does not close #14996 ([as per here](https://github.com/ray-project/ray/issues/14996#issuecomment-810538992)).

Good point! 
Will wrap the body of `_get_node_type` with
```
try:
except Exception:
    return "unknown"
```

**Comment by DmitriGekhtman** [comment]
> > @DmitriGekhtman This function (`def _terminate_nodes(self, node_ids):`) looks reasonable.
> > My concern is that this does not close #14996 ([as per here](https://github.com/ray-project/ray/issues/14996#issuecomment-810538992)).
> 
> Good point!
> Will wrap the body of `_get_node_type` with
> 
> ```
> try:
> except Exception:
>     return "unknown"
> ```

Hmm, also will have these messages printed only for the "active_nodes_to_terminate".

**Comment by DmitriGekhtman** [comment]
@AmeerHajAli Oh oops -- I now understand your question about max_workers. I added the new logic under the wrong `if nodes_to_terminate:`.

File: python/ray/autoscaler/_private/autoscaler.py [hunk 1]
```diff
@@ -267,18 +267,25 @@ def _update(self):
...
                         "Removing {} nodes of type " +
                         self._get_node_type(node_id) + " (launch failed).",
                         quantity=1,
                         aggregate=operator.add)
-                    nodes_to_terminate.append(node_id)
-                    self.num_failed_updates[node_id] += 1
-                del self.updaters[node_id]
-            if nodes_to_terminate:
-                self.provider.terminate_nodes(nodes_to_terminate)
+                self.provider.terminate_nodes(active_nodes_to_terminate)
```
Comment (DmitriGekhtman) [inline python/ray/autoscaler/_private/autoscaler.py:288]: The diff is hard to read -- I moved the `logger.error` / `self.event_summarizer` part into a for loop under the `if nodes_to_terminate` block.

File: python/ray/autoscaler/_private/autoscaler.py [hunk 2]
```diff
@@ -267,18 +267,25 @@ def _update(self):
...
+                    self.num_failed_updates[node_id] += 1
+                del self.updaters[node_id]
+            if nodes_to_terminate:
+                # Some nodes in nodes_to_terminate may have been terminated
+                # during an update. Only terminate currently active nodes.
+                active_nodes = self.workers()
+                active_nodes_to_terminate = [
+                    node for node in nodes_to_terminate if node in active_nodes
+                ]
+                for node_id in active_nodes_to_terminate:
```
Comment (ijrsvt) [inline python/ray/autoscaler/_private/autoscaler.py:280]: Should we debug log/info log on nodes that are in `nodes_to_terminate` and not in `active_nodes`?

Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/autoscaler.py]: Good idea.
That also makes things a bit more symmetric.

**Comment by DmitriGekhtman** [comment]
Will add unit tests for the `terminate_workers` function and the post-update processing logic.

**Review by AmeerHajAli — DISMISSED** [review]
Instead of doing that I suggest 2 solutions:
1) make terminating a terminated node OK (I don't see why that would be a problem).
2) take a look at my suggestion (lets start with this solution)

File: python/ray/autoscaler/_private/autoscaler.py [hunk 3]
```diff
@@ -267,18 +267,30 @@ def _update(self):
...
-                        quantity=1,
-                        aggregate=operator.add)
                     nodes_to_terminate.append(node_id)
                     self.num_failed_updates[node_id] += 1
                 del self.updaters[node_id]
             if nodes_to_terminate:
-                self.provider.terminate_nodes(nodes_to_terminate)
+                # Some nodes in nodes_to_terminate may have been terminated
+                # during an update. Only terminate currently active nodes.
+                active_nodes = self.workers()
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/autoscaler.py:276]:
```suggestion
                non_terminated_nodes = self.workers()
```

File: python/ray/autoscaler/_private/autoscaler.py [hunk 4]
```diff
@@ -267,18 +267,30 @@ def _update(self):
...
-                                 "failed to setup/initialize node.")
-                    self.event_summarizer.add(
-                        "Removing {} nodes of type " +
-                        self._get_node_type(node_id) + " (launch failed).",
-                        quantity=1,
-                        aggregate=operator.add)
                     nodes_to_terminate.append(node_id)
                     self.num_failed_updates[node_id] += 1
                 del self.updaters[node_id]
             if nodes_to_terminate:
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/autoscaler.py:273]:
```suggestion
            if failed_nodes:
```

File: python/ray/autoscaler/_private/autoscaler.py [hunk 5]
```diff
@@ -267,18 +267,30 @@ def _update(self):
...
-                self.provider.terminate_nodes(nodes_to_terminate)
+                # Some nodes in nodes_to_terminate may have been terminated
+                # during an update. Only terminate currently active nodes.
+                active_nodes = self.workers()
+                active_nodes_to_terminate = []
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/autoscaler.py:277]:
```suggestion
                nodes_to_terminate = []
```

File: python/ray/autoscaler/_private/autoscaler.py [hunk 6]
```diff
@@ -267,18 +267,30 @@ def _update(self):
...
+                for node_id in nodes_to_terminate:
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/autoscaler.py:278]:
```suggestion
                for node_id in failed_nodes:
```

File: python/ray/autoscaler/_private/autoscaler.py [hunk 7]
```diff
@@ -267,18 +267,30 @@ def _update(self):
...
+                    if node_id in active_nodes:
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/autoscaler.py:279]:
```suggestion
                    if node_id in non_terminated_nodes:
```

File: python/ray/autoscaler/_private/autoscaler.py (hunk 4 above)
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/autoscaler.py:273]:
```suggestion
            if failed_nodes:
```

File: python/ray/autoscaler/_private/autoscaler.py [hunk 8]
```diff
@@ -267,18 +267,30 @@ def _update(self):
...
+                            self._get_node_type(node_id) + " (launch failed).",
+                            quantity=1,
+                            aggregate=operator.add)
+                    else:
+                        logger.error(f"StandardAutoscaler: {node_id}: "
+                                     " Failed to setup/initialize node."
+                                     " Node already terminated.")
+                self.terminate_workers(active_nodes_to_terminate)
 
             nodes = self.workers()
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/autoscaler.py:295]:
```suggestion
```

File: python/ray/autoscaler/_private/autoscaler.py [hunk 9]
```diff
@@ -267,18 +267,30 @@ def _update(self):
...
+                        self.event_summarizer.add(
+                            "Removing {} nodes of type " +
+                            self._get_node_type(node_id) + " (launch failed).",
+                            quantity=1,
+                            aggregate=operator.add)
+                    else:
+                        logger.error(f"StandardAutoscaler: {node_id}: "
+                                     " Failed to setup/initialize node."
+                                     " Node already terminated.")
+                self.terminate_workers(active_nodes_to_terminate)
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/autoscaler.py:293]:
```suggestion
                # Some nodes in failed_nodes may have been terminated
                # during an update (for being idle after missing a heartbeat). 
                # Only terminate currently non terminated nodes.
                non_terminated_nodes = self.workers()
                nodes_to_terminate = []
                for node_id in failed_nodes:
                    if node_id in non_terminated_nodes:
                        nodes_to_terminate.append(node_id)
                        logger.error(f"StandardAutoscaler: {node_id}:"
                                     " Terminating. Failed to setup/initialize"
                                     " node.")
                        self.event_summarizer.add(
                            "Removing {} nodes of type " +
                            self._get_node_type(node_id) + " (launch failed).",
                            quantity=1,
                            aggregate=operator.add)
                if nodes_to_terminate:
                    self.provider.terminate_nodes(nodes_to_terminate)
                    for node in nodes_to_terminate:
                        self.node_tracker.untrack(node)
                    nodes = self.workers()
```

File: python/ray/autoscaler/_private/autoscaler.py [hunk 10]
```diff
@@ -309,6 +321,27 @@ def _update(self):
...
+            active_nodes = self.workers()
+            for node_id in node_ids:
+                if node_id in active_nodes:
+                    logger.error(f"StandardAutoscaler: {node_id}: "
+                                 " Trying again to terminate.")
+                    self.provider.terminate_node(node_id)
+                else:
+                    logger.error(f"StandardAutoscaler: {node_id}: "
+                                 " Node already terminated.")
+
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/autoscaler.py:344]:
```suggestion
```

File: python/ray/autoscaler/_private/autoscaler.py [hunk 11]
```diff
@@ -598,10 +631,13 @@ def recover_if_needed(self, node_id, now):
         self.updaters[node_id] = updater
 
     def _get_node_type(self, node_id: str) -> str:
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/autoscaler.py:633]: Keep the function the same

File: python/ray/autoscaler/_private/autoscaler.py [hunk 12]
```diff
@@ -731,7 +767,7 @@ def kill_workers(self):
         logger.error("StandardAutoscaler: kill_workers triggered")
         nodes = self.workers()
         if nodes:
-            self.provider.terminate_nodes(nodes)
+            self.terminate_workers(nodes)
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/autoscaler.py:770]:
```suggestion
```

**Comment by DmitriGekhtman** [comment]
> 1. make terminating a terminated node OK (I don't see why that would be a problem).

This would in any case have to be done with exception handling in the autoscaler, otherwise we would have to guarantee that none of the NodeProviders raise exceptions in this situation and add a requirement to the NodeProviderInterface that terminating the same node twice has to be handled there.

**Comment by AmeerHajAli** [comment]
@DmitriGekhtman , Done. Can you actually move everything under `if completed` to a new function `handle_completed_updaters` ?

**Comment by DmitriGekhtman** [comment]
@AmeerHajAli do you think we should do additional error handling in the autoscaler for nodes that are terminated twice?

The PR in its current state resolves one race condition leading to double termination.
This PR fixed another race condition leading to the same behavior:
https://github.com/ray-project/ray/pull/14976

I imagine problems of this sort could arise again, and it could help to be defensive.

File: python/ray/autoscaler/_private/autoscaler.py [hunk 13]
```diff
@@ -267,20 +268,32 @@ def _update(self):
...
+                    if node_id in non_terminated_nodes:
+                        nodes_to_terminate.append(node_id)
+                        logger.error(f"StandardAutoscaler: {node_id}:"
+                                     " Terminating. Failed to setup/initialize"
+                                     " node.")
+                        self.event_summarizer.add(
+                            "Removing {} nodes of type " +
+                            self._get_node_type(node_id) + " (launch failed).",
+                            quantity=1,
+                            aggregate=operator.add)
```
Comment (ijrsvt) [inline python/ray/autoscaler/_private/autoscaler.py:291]: Can we add an `else` statement here that debug/info logs double terminations?

File: python/ray/autoscaler/_private/autoscaler.py [hunk 14]
```diff
@@ -267,20 +268,32 @@ def _update(self):
...
+                if nodes_to_terminate:
+                    self.provider.terminate_nodes(nodes_to_terminate)
+                    for node in nodes_to_terminate:
+                        self.node_tracker.untrack(node)
```
Comment (ijrsvt) [inline python/ray/autoscaler/_private/autoscaler.py:295]: Was this missing in the previous implementation?

**Comment by ijrsvt** [comment]
> I imagine problems of this sort could arise again, and it could help to be defensive.

+1 on being defensive here. That being said, I think the current behavior (only terminating nodes that are in `nodes` is okay).

File: python/ray/autoscaler/_private/autoscaler.py (hunk 14 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/autoscaler.py:295]: I don't think so.

We do need to also un-track nodes which failed the update and were already terminated.

File: python/ray/autoscaler/_private/autoscaler.py (hunk 13 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/autoscaler.py:293]: Add an else with a warning.

**Review by ijrsvt — APPROVED** [review]
This looks good to me at this point!

**Comment by DmitriGekhtman** [comment]
>  Can you actually move everything under `if completed` to a new function `handle_completed_updaters` ?

@AmeerHajAli 
I think it might actually be better to do this refactor in another PR.
Also -- it's not immediately obvious how to do this cleanly, in particular how to update the variable `nodes`.
But for sure we should separate each step of `_update` that creates or deletes nodes into a separate function.

Did we open an issue a while ago to clean up autoscaler code?

**Comment by DmitriGekhtman** [comment]
Will do these things:
- Attempt a unit test for the new autoscaler logic
- Add a tiny bit of logic to k8s operator test to check the error handling on terminate_node

**Comment by AmeerHajAli** [comment]
I think it might be good sometimes to fail to see what kind of issues could happen. If we add all this logic and it ends up not ever happening, it becomes dead code.

**Comment by DmitriGekhtman** [comment]
> I think it might be good sometimes to fail to see what kind of issues could happen. If we add all this logic and it ends up not ever happening, it becomes dead code.

Yeah, I agree. There is some sort of balance to be achieved with error handling -- keep the system running reliably but don't swallow internal inconsistencies.

**Comment by DmitriGekhtman** [comment]
Added a unit test to validate the new completed-update-processing logic.
Unfortunately, it was interacting badly with the other tests in `test_autoscaler.py` so I had to make a new file `test_autoscaler2.py` for the new unit test.

Also added a test of the new K8s node provider exception handling.

**Comment by DmitriGekhtman** [comment]
ugh, new test is flaky

**Comment by AmeerHajAli** [comment]
> ugh, new test is flaky

What is going on?

**Comment by DmitriGekhtman** [comment]
Test should work now.

**Comment by DmitriGekhtman** [comment]
Test failures are unrelated, will rebase and run once more just for good measure.

**Comment by DmitriGekhtman** [comment]
good to merge

**Comment by DmitriGekhtman** [comment]
windows test_scheduling is known flaky.
@AmeerHajAli or @ijrsvt , could you merge this?

**Pull request merged**